wolfssl/wolfcrypt/lms.h

```diff
@@ -0,0 +1,131 @@
+                                     size_t len_private_key, void *context);
+typedef bool (*read_private_key_cb)(unsigned char *private_key,
+                                    size_t len_private_key, void *context);
+#else
+/* key callbacks */
+typedef int (*write_private_key_cb)(unsigned char *private_key,
+                                    size_t len_private_key, void *context);
+typedef int (*read_private_key_cb)(unsigned char *private_key,
+                                   size_t len_private_key, void *context);
+#endif /* if defined(HAVE_LIBLMS) */
+
+/* LMS/HSS signatures are defined by 3 parameters:
+ *   levels: number of levels of Merkle trees.
+ *   height: height of an individual Merkle tree.
+ *   winternitz: number of bits from hash used in a Winternitz chain.
+ *
+ * The acceptable parameter values are those in RFC8554:
+ *   levels = {1..8}
+ *   height = {5, 10, 15, 20, 25}
+ *   winternitz = {1, 2, 4, 8}
+ *
+ * The number of available signatures is:
+ *   N = 2 ** (levels * height)
+ *
+ * Signature sizes are determined by levels and winternitz
+ * parameters primarily, and height to a lesser extent:
+ *   - Larger levels values increase signature size significantly.
+ *   - Larger height values increase signature size moderately.
+ *   - Larger winternitz values will reduce the signature size, at
+ *     the expense of longer key generation and sign/verify times.
+ *
+ * Key generation time is strongly determined by the height of
+ * the first level tree. A 3 level, 5 height tree is much faster
+ * than 1 level, 15 height at initial key gen, even if the number
+ * of available signatures is the same.
+ * */
+
+/* Predefined LMS/HSS parameter sets for convenience. */
+enum wc_LmsParm {
+    WC_LMS_PARM_NONE      =  0,
+    WC_LMS_PARM_L1_H15_W2 =  1, /* 1 level Merkle tree of 15 height. */
+    WC_LMS_PARM_L1_H15_W4 =  2,
+    WC_LMS_PARM_L2_H10_W2 =  3, /* 2 level Merkle tree of 10 height. */
+    WC_LMS_PARM_L2_H10_W4 =  4,
+    WC_LMS_PARM_L2_H10_W8 =  5,
+    WC_LMS_PARM_L3_H5_W2  =  6, /* 3 level Merkle tree of 5 height. */
+    WC_LMS_PARM_L3_H5_W4  =  7,
+    WC_LMS_PARM_L3_H5_W8  =  8,
+    WC_LMS_PARM_L3_H10_W4 =  9, /* 3 level Merkle tree of 10 height. */
+    WC_LMS_PARM_L4_H5_W8  = 10, /* 4 level Merkle tree of 5 height. */
+};
+
+enum wc_LmsState {
+    WC_LMS_STATE_OK         = 0,
+    WC_LMS_STATE_NOT_INITED = 1,
+    WC_LMS_STATE_INITED     = 2,
+    WC_LMS_STATE_BAD        = 3, /* Can't guarantee key's state. */
+    WC_LMS_STATE_NOSIGS     = 4, /* Signatures exhausted. */
+};
+
+#ifdef __cplusplus
+    extern "C" {
+#endif
+/* Need an update_key_cb setting function.
+ * Also maybe a generate rand cb setter. */
+WOLFSSL_API int  wc_LmsKey_Init(LmsKey * key, enum wc_LmsParm lmsParm);
+WOLFSSL_API int  wc_LmsKey_Init_ex(LmsKey * key, int levels,
+    int height, int winternitz, void* heap, int devId);
+WOLFSSL_API int  wc_LmsKey_SetWriteCb(LmsKey * key, write_private_key_cb wf);
```

Comment:
Why `wf` ?

Reply:
wf == write function
rf == read function

Reply:
can you change to write_cb and read_cb?